IdentityCore/src/cache/crypto/MSIDAssymetricKeyKeychainGenerator.m

```diff
@@ -0,0 +1,276 @@
+    }
+    
+    // Add team prefix to keychain group if it is missed.
+    if (![keychainGroup hasPrefix:keychainUtil.teamId])
+    {
+        keychainGroup = [keychainUtil accessGroup:keychainGroup];
+    }
+    
+    _keychainGroup = keychainGroup;
+    
+    if (!_keychainGroup)
+    {
+        if (error)
+        {
+            *error = MSIDCreateError(MSIDErrorDomain, MSIDErrorInternal, @"Failed to set keychain access group.", nil, nil, nil, nil, nil, YES);
+        }
+        
+        MSID_LOG_WITH_CTX(MSIDLogLevelError, nil, @"Failed to set keychain access group.");
+        return nil;
+    }
+    
+    NSMutableDictionary *defaultKeychainQuery = [@{(id)kSecAttrAccessGroup : self.keychainGroup} mutableCopy];
+    [defaultKeychainQuery addEntriesFromDictionary:[self additionalPlatformKeychainAttributes]];
+    
+    self.defaultKeychainQuery = defaultKeychainQuery;
+    return self;
+}
+
+#pragma mark - MSIDAssymetricKeyGenerating
+
+- (MSIDAssymetricKeyPair *)generateKeyPairForAttributes:(MSIDAssymetricKeyLookupAttributes *)attributes
+                                                  error:(NSError **)error
+{
+    if ([NSString msidIsStringNilOrBlank:attributes.privateKeyIdentifier]
+        || [NSString msidIsStringNilOrBlank:attributes.publicKeyIdentifier])
+    {
+        [self logAndFillError:@"Invalid key generation attributes provided" status:-1 error:error];
+        return nil;
+    }
+    
+    // 0. Cleanup any previous state
+    BOOL cleanupResult = [self cleanupKeychainForAttributes:attributes error:error];
+    
+    if (!cleanupResult)
+    {
+        [self logAndFillError:@"Failed to cleanup keychain prior to generating new keypair. Proceeding might result in unexpected results. Keychain might need to be manually cleaned to recover." status:-1 error:error];
+        return nil;
+    }
+    
+    // 1. Generate keypair
+    NSDictionary *keyPairAttr = [self keychainQueryWithAttributes:[attributes assymetricKeyPairAttributes]];
+    
+    SecKeyRef publicKeyRef = NULL;
+    SecKeyRef privateKeyRef = NULL;
+    OSStatus status = SecKeyGeneratePair((__bridge CFDictionaryRef)keyPairAttr, &publicKeyRef, &privateKeyRef);
+    
+    if (status != errSecSuccess)
+    {
+        [self logAndFillError:@"Failed to generate keypair" status:status error:error];
+        return nil;
+    }
+    
+    // 2. Return keys
+    MSIDAssymetricKeyPair *keyPair = [[MSIDAssymetricKeyPair alloc] initWithPrivateKey:privateKeyRef publicKey:publicKeyRef];
+    
+    if (privateKeyRef) CFRelease(privateKeyRef);
+    if (publicKeyRef) CFRelease(publicKeyRef);
+    
+    return keyPair;
+}
+
+- (MSIDAssymetricKeyPair *)readOrGenerateKeyPairForAttributes:(MSIDAssymetricKeyLookupAttributes *)attributes
+                                                        error:(NSError **)error
+{
+    NSError *readError = nil;
+    MSIDAssymetricKeyPair *keyPair = [self readKeyPairForAttributes:attributes error:&readError];
+    
+    if (keyPair || readError)
```

Comment:
Can we change this logic to either have the keyPair or readError. For instance, the function either returns the object or fills the error.

Reply:
@rohitnarula7176, this code will return early if it encountered an error or found a key in cache. That is why both items are checked. I don't think it's possible to change it to only check one. We can have two if conditions if that helps with readability.

Reply:
Agreed to follow up separately.